src/aleph/http/http2.clj

```diff
@@ -0,0 +1,503 @@
+     equal to 'trailers'."
+    [^Http2Headers h2-headers ^String header-name header-value]
+    (log/debug "Adding HTTP header" header-name ": " header-value)
+
+    (if (nil? header-name)
+      (throw (IllegalArgumentException. "Header name cannot be nil"))
+      (let [header-name (str/lower-case header-name)]       ; http2 requires lowercase headers
+        (cond
+          (nil? header-value)
+          (illegal-arg (str "Invalid nil value for header '" header-name "'"))
+
+          (invalid-headers header-name)
+          (illegal-arg (str "Forbidden HTTP/2 header: \"" header-name "\""))
+
+          ;; See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding
+          (and (.equals "transfer-encoding" header-name)
+               (not (.equals "trailers" header-value)))
+          (illegal-arg "Invalid value for 'transfer-encoding' header. Only 'trailers' is allowed.")
+
+          (sequential? header-value)
+          (.add h2-headers ^CharSequence header-name ^Iterable header-value)
+
+          :else
+          (.add h2-headers ^CharSequence header-name ^Object header-value))))))
+
+(defn- ring-map->netty-http2-headers
+  "Builds a Netty Http2Headers object from a Ring map."
+  ^DefaultHttp2Headers
+  [req]
+  (prn req)
+  (let [headers (get req :headers)
+        h2-headers (doto (DefaultHttp2Headers.)
+                         (.method (-> req (get :request-method) name str/upper-case))
+                         (.scheme (-> req (get :scheme) name))
+                         (.authority (:authority req))
+                         (.path (str (get req :uri)
+                                     (when-let [q (get req :query-string)]
+                                       (str "?" q)))))]
+    (when (:status req)
+      (.status h2-headers (-> req (get :status) str)))
+
+    (when headers
+      (run! #(add-header h2-headers (key %) (val %))
+            headers))
+    h2-headers))
+
+(defn netty-http2-headers->map
+  "Returns a map of Ring headers from a Netty Http2Headers object.
+
+   Includes pseudo-headers."
+  [^Http2Headers headers]
+  (loop [iter (.iterator headers)
+         result {}]
+    (if (.hasNext iter)
+      (let [entry (.next iter)]
+        (recur iter
+               (assoc result
+                      (.toString ^CharSequence (key entry))
+                      (.toString ^CharSequence (val entry)))))
+      result)))
+
+(defn- try-set-content-length!
+  "Attempts to set the `content-length` header if not already set.
+
+   Will skip if it's a response and the status code is 1xx or 204.
+   Negative length values are ignored."
+  ^Http2Headers
+  [^Http2Headers headers ^long length]
+  (when-not (.get headers "content-length")
+    (if (some? (.status headers))
+      ;; TODO: consider switching to netty's HttpResponseStatus and HttpStatusClass for clarity
+      (let [code (-> headers (.status) (Long/parseLong))]
+        (when-not (or (<= 100 code 199)
```

Comment:
Good catch, but sadly, no. The primitive-math fns don't support arities higher than 2. I thought about using two `<=` calls, but thought it would look too weird. OTOH, if this is about performance, I _should_ do it...

Reply:
OK, converted to use primitive-math, too